src/collapse/dtrace.rs

```diff
@@ -1,54 +1,83 @@
-use super::util::fix_partially_demangled_rust_symbol;
-use super::Collapse;
-use fnv::FnvHashMap;
-use log::warn;
 use std::borrow::Cow;
 use std::collections::VecDeque;
-use std::io;
-use std::io::prelude::*;
+use std::io::{self, prelude::*};
 
-/// Settings that change how frames are named from the incoming stack traces.
-///
-/// All options default to off.
-#[derive(Clone, Debug, Default)]
+use log::warn;
+
+use crate::collapse::common::{self, CollapsePrivate, Occurrences};
+
+/// `dtrace` folder configuration options.
+#[derive(Clone, Debug)]
 pub struct Options {
-    /// include function offset (except leafs)
+    /// Demangle function names. Default is `false`.
+    pub demangle: bool,
+
+    /// Include function offset (except leafs). Default is `false`.
     pub includeoffset: bool,
 
-    /// Demangle function names
-    pub demangle: bool,
+    /// The number of threads to use. Default is the number of logical cores on your machine.
+    pub nthreads: usize,
+}
+
+impl Default for Options {
+    fn default() -> Self {
+        Self {
+            demangle: false,
+            includeoffset: false,
+            nthreads: *common::DEFAULT_NTHREADS,
+        }
+    }
 }
 
 /// A stack collapser for the output of dtrace `ustrace()`.
 ///
 /// To construct one, either use `dtrace::Folder::default()` or create an [`Options`] and use
 /// `dtrace::Folder::from(options)`.
-#[derive(Default)]
+#[derive(Clone)]
```

Comment:
It's arguably a little weird for this to be `derive(Clone)`, because cloning while there's, say, stuff in `stack`, will lead to counter-intuitive results. May be this implementation has to be manual to ensure that the new `Folder` is "clean".

Reply:
I actually have the opposite intuition. Since `Clone` is so very rarely implemented manually in rust, my first instinct when I call `clone` on something is that I get an exact copy, internal state and all. I would be pretty surprised if it did anything else.

That said, from the users perspective, they don't (and shouldn't) know anything about internal state; so I guess the question is what would they expect if they ever want to call `clone`. I'm struggling to think of why they would, but my lack of imagination doesn't mean they won't want to for something. 

Can you think of an instance where the user would want to call `clone` and the `[derive(Clone)]` version would trip them up?

Reply:
Hmm, that's a good point. It makes me think that we might actually _not_ want to implement `Clone` here, and instead have a private function that gives you a clone. What do you think?

Reply:
Well, from the user's perspective, I'm not sure there's any downside of keeping `[derive(Clone)]` since they should never be able to get a `Folder` into a weird state. They can only call `collapse`, `collapse_file` and `is_applicable` on a `Collapse` and all of those methods are guaranteed to leave the `Folder` in a clean state once they're called. So I'm not sure allowing them to `clone` with the `[derive(Clone)]` has any downside.

Doing a private method, on the other hand, has two downsides:

* Maybe the user will need to clone for some reason we can't think of (I admit I can't think of why they would want to, but this might just be a function of my poor imagination).

* More importantly, a private method would have to be an unprovided method on `CollapsePrivate` because `clone` is currently used in `collapse_multithreaded`. This means we'd be asking implementors of `CollapsePrivate` to implement yet another required method, which is a downside, it seems.

Reply:
I went ahead and dropped the `Clone` derive. Let me know if you change your mind. Otherwise, I think that's the last open issue.

Reply:
In this context I'd rather err on the side of not exposing a method to the user that _may_ give them weird results if they get a handle to it at the wrong time. The change (which looks good by the way) also makes the requirements of the cloning method explicit, which I like. From a user's perspective, the `Options` are generally `Clone`, so the trick there will just be to clone the options instead if you're going to be making multiple folders.

Reply:
In this context I'd rather err on the side of not exposing a method to the user that _may_ give them weird results if they get a handle to it at the wrong time. The change (which looks good by the way) also makes the requirements of the cloning method explicit, which I like. From a user's perspective, the `Options` are generally `Clone`, so the trick there will just be to clone the options instead if you're going to be making multiple folders.